### ISSUE TITLE
missing main() in C++ tests

### ISSUE THREAD

**Issue by sharkcz** [msgpack/msgpack-c#970]
When building the `cpp_master` branch in Fedora 34 (and Rawhide) on x86_64 (or ppc64le) with boost 1.75 (and 1.76) the tests can't be linked because `main()` is missing.
````
[dan@talos msgpack-c]$ git clean -xfd && LANG=C cmake -S . -B build -DMSGPACK_BUILD_TESTS=ON -DMSGPACK_CXX20=ON && LANG=C make -C build                                                   :cpp
Removing build/
-- The CXX compiler identification is GNU 11.2.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Boost: /usr/include (found version "1.75.0")  
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.11") 
-- Found Boost: /usr/include (found version "1.75.0") found components: unit_test_framework system 
-- Found Doxygen: /usr/bin/doxygen (found version "1.9.1") found components: doxygen dot 
-- Configuring done
-- Generating done
-- Build files have been written to: /opt2/src/msgpack-c/build
make: Entering directory '/opt2/src/msgpack-c/build'
make[1]: Entering directory '/opt2/src/msgpack-c/build'
make[2]: Entering directory '/opt2/src/msgpack-c/build'
make[2]: Leaving directory '/opt2/src/msgpack-c/build'
make[2]: Entering directory '/opt2/src/msgpack-c/build'
[  1%] Building CXX object test/CMakeFiles/multi_file.dir/multi_file1.cpp.o
[  2%] Building CXX object test/CMakeFiles/multi_file.dir/multi_file2.cpp.o
[  3%] Linking CXX executable multi_file
make[2]: Leaving directory '/opt2/src/msgpack-c/build'
[  3%] Built target multi_file
make[2]: Entering directory '/opt2/src/msgpack-c/build'
make[2]: Leaving directory '/opt2/src/msgpack-c/build'
make[2]: Entering directory '/opt2/src/msgpack-c/build'
[  4%] Building CXX object test/CMakeFiles/msgpack_cpp20.dir/msgpack_cpp20.cpp.o
[  6%] Linking CXX executable msgpack_cpp20
/usr/bin/ld: /usr/lib/gcc/ppc64le-redhat-linux/11/../../../../lib64/crt1.o:(.rodata+0x8): undefined reference to `main'
collect2: error: ld returned 1 exit status
make[2]: *** [test/CMakeFiles/msgpack_cpp20.dir/build.make:100: test/msgpack_cpp20] Error 1
make[2]: Leaving directory '/opt2/src/msgpack-c/build'
make[1]: *** [CMakeFiles/Makefile2:985: test/CMakeFiles/msgpack_cpp20.dir/all] Error 2
make[1]: Leaving directory '/opt2/src/msgpack-c/build'
make: *** [Makefile:146: all] Error 2
make: Leaving directory '/opt2/src/msgpack-c/build'
````
I have tracked the issue to the header used for the Boost `test` module. When I switch from `boost/test/unit_test.hpp` to `boost/test/included/unit_test.hpp` as suggested by the Boost docs (https://www.boost.org/doc/libs/1_77_0/libs/test/doc/html/boost_test/tests_organization/test_tree/master_test_suite.html#ref_BOOST_TEST_MODULE), then the tests can be linked correctly. Not sure what is wrong, because your CI shows the tests building OK ...

**Comment by redboltz** [comment]
@kovdan01 , could you investigate the issue?

**Comment by kovdan01** [comment]
@redboltz yes, I'll try to do it ASAP

**Comment by redboltz** [comment]
Perhaps `BOOST_TEST_DYN_LINK` compiler flag is needed.
My other project do that.
https://github.com/redboltz/mqtt_cpp/blob/39e2c4cd60266ff8d52157d068626b75686b1967/test/system/CMakeLists.txt#L84

**Comment by redboltz** [comment]
@sharkcz the link error is not reproduced on my local environment (and CI).
In order to investigate the error, could you add cmake option `-DCMAKE_CXX_FLAGS="-DBOOST_TEST_DYN_LINK"` ?

**Comment by sharkcz** [comment]
> @sharkcz the link error is not reproduced on my local environment (and CI).
> In order to investigate the error, could you add cmake option `-DCMAKE_CXX_FLAGS="-DBOOST_TEST_DYN_LINK"` ?

yes, adding `-DBOOST_TEST_DYN_LINK` to the compiler flags fixes the issue

**Comment by redboltz** [comment]
@sharkcz thanks. Please use the flag as a temporary solution.

@kovdan01 , `BOOST_TEST_DYN_LINK` should be automatically defined by default. Could you add the flag to CMakeLists.txt ?
As I mentioned https://github.com/msgpack/msgpack-c/issues/970#issuecomment-908361746 , if you support static link boost then more mechanism is required. I personally think that supporting only dynamic link boost is good enough.

**Comment by kovdan01** [comment]
@redboltz OK, I'll submit a PR with the fix

**Comment by kovdan01** [comment]
@redboltz Submitted #971

**Comment by redboltz** [comment]
Fixed by #971.
